src/MultiFormats/Registry/HashingAlgorithm.cs

```diff
@@ -0,0 +1,311 @@
+        /// </value>
+        public string Name { get; private set; }
+
+        /// <summary>
+        ///   The IPFS number assigned to the hashing algorithm.
+        /// </summary>
+        /// <value>
+        ///   Valid hash codes at <see href="https://github.com/multiformats/multicodec/blob/master/table.csv">hashtable.csv</see>.
+        /// </value>
+        public int Code { get; private set; }
+
+        /// <summary>
+        ///   The size, in bytes, of the digest value.
+        /// </summary>
+        /// <value>
+        ///   The digest value size in bytes. Zero indicates that the digest
+        ///   is non fixed.
+        /// </value>
+        public int DigestSize { get; private set; }
+
+        /// <summary>
+        ///   Returns a cryptographic hash algorithm that can compute
+        ///   a hash (digest).
+        /// </summary>
+        public Func<HashAlgorithm> Hasher { get; private set; }
+
+        /// <summary>
+        ///   Use <see cref="Register"/> to create a new instance of a <see cref="HashingAlgorithm"/>.
+        /// </summary>
+        private HashingAlgorithm() { }
+
+        /// <summary>
+        ///   The <see cref="Name"/> of the hashing algorithm.
+        /// </summary>
+        public override string ToString() { return Name; }
+
+        /// <summary>
+        ///   Register a new IPFS hashing algorithm.
+        /// </summary>
+        /// <param name="name">
+        ///   The name of the algorithm.
+        /// </param>
+        /// <param name="code">
+        ///   The IPFS number assigned to the hashing algorithm.
+        /// </param>
+        /// <param name="digestSize">
+        ///   The size, in bytes, of the digest value.
+        /// </param>
+        /// <param name="hasher">
+        ///   A <c>Func</c> that returns a <see cref="HashAlgorithm"/>.  If not specified, then a <c>Func</c> is created to
+        ///   throw a <see cref="NotImplementedException"/>.
+        /// </param>
+        /// <returns>
+        ///   A new <see cref="HashingAlgorithm"/>.
+        /// </returns>
+        public static HashingAlgorithm Register(string name,
+            int code,
+            int digestSize,
+            Func<HashAlgorithm> hasher = null)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            
+            if (Names.ContainsKey(name))
+            {
+                throw new ArgumentException(string.Format("The IPFS hashing algorithm '{0}' is already defined.",
+                    name));
+            }
+            
+            if (Codes.ContainsKey(code))
+            {
+                throw new ArgumentException(
+                    $"The IPFS hashing algorithm code 0x{code:x2} is already defined.");
+            }
+            
+            if (hasher == null)
+            {
+                hasher = () => throw new NotImplementedException(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Introduce a new variable instead of reusing the parameter 'hasher'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)